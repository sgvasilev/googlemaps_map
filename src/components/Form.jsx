import "../App.css"

import axios from "axios"
import React, { useState, useContext, useEffect, useCallback } from "react"
import { v4 as uuidv4 } from "uuid"
import debounce from "lodash.debounce"
import { RiRouteLine } from "react-icons/ri"
import styled from "styled-components"
//import data from "../components/MapLoader"

import userMarkerContext from "../context/userMarkers"
import MarkersList from "./MarkersList"

const Form = () => {
  const thisInput = React.createRef()

  const { setMarker, marker, userMarkers, addMarker, setUserMarkers } =
    useContext(userMarkerContext)

  const _handleSubmit = (e) => {
    e.preventDefault()
    if (marker.toString().trim() === "" || marker.toString().trim() === null)
      return
    else {
      const newMarker = {
        id: uuidv4(),
        place: marker,
      }
      addMarker(newMarker)
      setMarker("")
      thisInput.current.value = ""
    }
  }
  console.log("form render")
  const _handleChange = (newPlace) => {
    setMarker(newPlace)
  }
  const _handleDeb = useCallback(
    debounce(() => {
      //console.log(data)
    }, 900),
    []
  )

  return (
    <div>
      <form onSubmit={_handleSubmit} className="userrouting__form">
        <label
          style={{ display: "flex", flexDirection: "column", width: "300px" }}
        >
          Куда поедем?
          <input
            onChange={(e) => {
              _handleChange(e.target.value)
              _handleDeb()
            }}
            placeholder="Новая точка маршрута"
            className="userrouting__input"
            type="text"
            name="name"
            //value={marker}
            ref={thisInput}
            id="autocomplete"
          />
        </label>

        <button
          style={{
            background: "transparent",
            border: "none",
            cursor: "pointer",
          }}
          type="submit"
        >
          <RiRouteLine style={{ position: "absolute" }} size={24} />
        </button>
      </form>
      <MarkersList />
    </div>
  )
}

export default Form
