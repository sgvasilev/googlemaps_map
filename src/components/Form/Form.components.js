import styled from "styled-components"
export const FindRouteButton = styled.button`
  display: flex;
  color: inherit;
  cursor: pointer;
  background: transparent;
  border: none;
  width: 40px;
  height: 40px;
  align-items: center;
  margin-top: 2px;
`
export const FormLabel = styled.label`
  display: flex;
  flex-direction: column;
  width: 300px;
  margin-bottom: 20px;
`
export const RouteSuggestionsStyleFix = styled.div`
  border: 2px solid rgb(200, 220, 200);
  border-top: none;
`
