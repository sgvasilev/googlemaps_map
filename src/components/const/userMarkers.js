export const ADD_MARKER = "ADD_MARKER"
export const INITIAL_STATE = "INITIAL_STATE"
export const DELETE_MARKER = "DELETE_MARKER"
export const REORDER_MARKERS = "REORDER_MARKERS"
export const MARKER_LIST = "MARKER_LIST"
export const RENEW_MARKER = "RENEW_MARKER"
