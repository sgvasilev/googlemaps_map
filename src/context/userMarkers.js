import { createContext, useReducer, useState } from "react"
import markersReducer from "../reducers/markersReducer"
const userMarkerContext = createContext()

export const UserMarkersProvider = ({ children }) => {
  const initialState = {
    userMarkers: [],
    marker: [],
  }
  const addMarker = (data) => {
    dispatch({
      type: "ADD_MARKER",
      payload: data,
    })
  }
  const setMarker = (marker) => {
    dispatch({
      type: "MARKER_LIST",
      payload: marker,
    })
  }

  const deleteMarker = (id) => {
    const data = state.userMarkers.filter((el) => el.id !== id)
    console.log("data", data)
    dispatch({
      type: "DELETE_MARKER",
      payload: data,
    })
  }
  const reorderMarkers = (items) => {
    dispatch({
      type: "REORDER_MARKERS",
      payload: items,
    })
  }

  const renewMarker = (id, latLng, newLocation) => {
    const tempData = state.userMarkers.find((marker) => marker.id === id)
    tempData.position = latLng
    if ((newLocation !== null) & (newLocation !== undefined)) {
      tempData.description = newLocation
    }
    const tempIndex = state.userMarkers.findIndex((el) => el.id === id)
    const tempUserMarkers = [...state.userMarkers]
    tempUserMarkers.splice(tempIndex, 1)
    tempUserMarkers.splice(tempIndex, 0, tempData)
    dispatch({
      type: "RENEW_MARKER",
      payload: tempUserMarkers,
    })
  }
  const [map, setMap] = useState()
  const [state, dispatch] = useReducer(markersReducer, initialState)
  return (
    <userMarkerContext.Provider
      value={{
        deleteMarker,
        addMarker,
        reorderMarkers,
        userMarkers: state.userMarkers,
        setMap,
        map,
        marker: state.marker,
        setMarker,
        renewMarker,
      }}
    >
      {children}
    </userMarkerContext.Provider>
  )
}

export default userMarkerContext
