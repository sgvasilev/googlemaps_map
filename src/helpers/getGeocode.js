export const getGeocode = async (input) => {
  const geocoder = new window.google.maps.Geocoder()
  try {
    // const result = await geocoder.geocode({ location: input })
    // const data = result.results[0].formatted_address
    const data = "Moved"
    return data
  } catch (error) {
    throw new Error("Невозможно найти выбраный Вами адресс")
  }
}
