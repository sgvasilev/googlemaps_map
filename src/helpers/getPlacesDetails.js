export const getPlacesDetails = async (place, map) => {
  const mapRef = map
  const request = {
    placeId: place.place_id,
    fields: ["name", "formatted_address", "place_id", "geometry"],
  }

  let data = new Promise((resolve, reject) => {
    try {
      new window.google.maps.places.PlacesService(mapRef).getDetails(
        request,
        (place) => resolve(place?.geometry?.location)
      )
    } catch (e) {
      console.log(e, "Ooops")
      reject(e)
    }
  })
  return data
}
