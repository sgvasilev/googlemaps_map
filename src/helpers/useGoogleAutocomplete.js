import React, { useState, useEffect } from "react"
import { googleAutocompleteDeb } from "./googleAutocompleteDeb"

function useGoogleAutocomplete(text = "", debounce = 300) {
  const [predictions, setPredictions] = useState([])
  useEffect(() => {
    const handleDebounce = setTimeout(async () => {
      try {
        if (!text) return
        let data = await googleAutocompleteDeb(text)
        setPredictions(data)
      } catch (e) {
        return e
      }
    }, debounce)

    return () => clearTimeout(handleDebounce)
  }, [text, debounce])
  return predictions
}
export { useGoogleAutocomplete }
