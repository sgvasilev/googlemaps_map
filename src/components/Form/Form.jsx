import "../../App.css"

import React, { useState, useContext } from "react"

import { RiRouteLine as FindRouteIcon } from "react-icons/ri"
import {
  FindRouteButton,
  FormLabel,
  RouteSuggestionsStyleFix,
} from "./Form.components"

import userMarkerContext from "../../context/userMarkers"
import MarkersList from "../MarkersList"
import { useGoogleAutocomplete } from "../../helpers/useGoogleAutocomplete"
import RouteInput from "../RouteInput"
import RouteSuggestions from "../RouteSuggestions"
import { getPlacesDetails } from "../../helpers/getPlacesDetails"

const Form = () => {
  const { addMarker, userMarkers, map } = useContext(userMarkerContext)

  const [searchValue, setSearchValue] = useState("")
  const predictions = useGoogleAutocomplete(searchValue)
  const thisInput = React.createRef()
  const [idlat, setIdLat] = useState({ lat: 55.7532, lng: 37.6204, id: "a" })
  const handleChange = (newPlace) => {
    setSearchValue(newPlace)
  }

  const addMockMarker = (e) => {
    e.preventDefault()
    addMarker({
      id: idlat.id,
      place: `mock data + ${idlat.id}`,
      description: "mock data",
      position: { lat: idlat.lat, lng: idlat.lng },
    })

    setIdLat((prev) => ({
      id: prev.id + "f",
      lat: Number(prev.lat) + 0.002,
      lng: prev.lng + 0.002,
    }))
    setSearchValue("")
  }
  const handeChoseSelection = (e, prediction) => {
    e.preventDefault()
    let latLngResult = getPlacesDetails(prediction, map)
    latLngResult.then((result) => {
      const latLng = { lat: result.lat(), lng: result.lng() }
      const existingMarker = userMarkers.filter(
        (marker) => marker.id === prediction.place_id
      )
      if (existingMarker.length >= 1) return
      addMarker({
        id: prediction.place_id,
        place: prediction.structured_formatting.main_text,
        description: prediction.description,
        position: latLng,
      })
      setSearchValue("")
    })
    thisInput.current.value = ""
  }

  return (
    <div style={{ alignItems: "center" }}>
      <form className="userrouting__form">
        <FormLabel>
          <button
            style={{
              width: "fit-content",
              margin: "0 auto",
              borderRadius: "8px",
              height: "50px",
              cursor: " pointer",
              marginBottom: "16px",
            }}
            onClick={addMockMarker}
          >
            add mock lat lng
          </button>
          <RouteInput
            value={searchValue}
            handleChange={handleChange}
            thisInput={thisInput}
          />
          {searchValue &&
            predictions?.map((prediction) => (
              <RouteSuggestionsStyleFix
                key={prediction.place_id}
                className="userrouting__form_stylefix"
              >
                <RouteSuggestions
                  handeChoseSelection={handeChoseSelection}
                  prediction={prediction}
                />
              </RouteSuggestionsStyleFix>
            ))}
        </FormLabel>
        {/* <FindRouteButton type="submit">
          <FindRouteIcon size={28} />
        </FindRouteButton> */}
      </form>
      <MarkersList />
    </div>
  )
}

export default Form

//   {
//     description: "Боровск военкомат, Ленина площадь, Боровск, Калужская область, Россия"
// matched_substrings: [{…}]
// place_id: "ChIJMQpdkvvPykYRBazMcjH8g34"
// reference: "ChIJMQpdkvvPykYRBazMcjH8g34"
// structured_formatting: {main_text: 'Боровск военкомат', main_text_matched_substrings: Array(1), secondary_text: 'Ленина площадь, Боровск, Калужская область, Россия'}
// terms: (5) [{…}, {…}, {…}, {…}, {…}]
// types: (2) ['point_of_interest', 'establishment']
//   }
