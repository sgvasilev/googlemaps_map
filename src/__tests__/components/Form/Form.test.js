import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
} from "@testing-library/react"
import renderer from "react-test-renderer"

import { useGoogleAutocomplete } from "../../../helpers/useGoogleAutocomplete"
import Form from "../../../components/Form/Form"
import { UserMarkersProvider } from "../../../context/userMarkers"
import { getPlacesDetails } from "../../../helpers/getPlacesDetails"

jest.mock("../../../helpers/useGoogleAutocomplete")
jest.mock("../../../helpers/getPlacesDetails")

afterEach(() => {
  cleanup()
})

describe("Should properly render Form component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <UserMarkersProvider>
          <Form />
        </UserMarkersProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("should properly handle search value", () => {
    render(
      <UserMarkersProvider>
        <Form />
      </UserMarkersProvider>
    )
    const userLocationInput = screen.getByTestId("input")
    expect(userLocationInput).toBeInTheDocument()
    fireEvent.change(userLocationInput, { target: { value: "Moscow" } })
    expect(userLocationInput.value).toBe("Moscow")
  })
  it("should properly show search results", async () => {
    render(
      <UserMarkersProvider>
        <Form />
      </UserMarkersProvider>
    )
    let latLng
    const predictions = [
      {
        description: "data test 1",
        place_id: 123,
        structured_formatting: {
          main_text: "Moscow data",
        },
        position: latLng,
      },
    ]

    getPlacesDetails.mockImplementation(() =>
      Promise.resolve({
        lat: () => 39,
        lng: () => 39,
      })
    )

    await waitFor(() => useGoogleAutocomplete.mockReturnValueOnce(predictions))
    expect(useGoogleAutocomplete).toBeCalledTimes(1)
    const userLocationInput = screen.getByTestId("input")
    fireEvent.change(userLocationInput, { target: { value: "Moscow" } })
    expect(useGoogleAutocomplete).toBeCalledTimes(2)
    const testData = await screen.findByText("data test 1")
    expect(testData).toBeInTheDocument()
    const userSuggestions = screen.getByTestId("user_suggestions")
    expect(userSuggestions).toBeInTheDocument()
    expect(userSuggestions).toHaveTextContent("data test 1")
    fireEvent.click(userSuggestions)
    await waitFor(() => expect(getPlacesDetails).toHaveBeenCalledTimes(1))
    expect(userLocationInput).toHaveTextContent("")
  })
})
