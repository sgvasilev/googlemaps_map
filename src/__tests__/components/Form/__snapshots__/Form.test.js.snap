// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should properly render Form component renders correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <form
    className="userrouting__form"
  >
    <label
      className="sc-gsDKAQ bHVlRh"
    >
      <button
        onClick={[Function]}
      >
        add mock
      </button>
      <input
        className="userrouting__input"
        data-testid="input"
        id="autocomplete"
        onChange={[Function]}
        placeholder="Новая точка маршрута"
        type="text"
      />
      
    </label>
  </form>
  <ul
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="markers"
    style={
      Object {
        "listStyle": "none",
        "paddingInlineStart": "8px",
      }
    }
  />
</div>
`;
